tasks.py

```diff
@@ -209,3 +209,84 @@ def get_markdown_spec(ctx):
         "https://raw.githubusercontent.com/Materials-Consortia/OPTIMADE/develop/optimade.rst "
         "> optimade.md"
     )
+
+
+@task(
+    help={
+        "pre-clean": "Remove the 'api_reference' sub directory prior to (re)creation."
+    }
+)
+def create_api_reference_docs(_, pre_clean=False):
+    """Create the API Reference in the documentation"""
+    import shutil
+
+    def write_file(full_path: Path, content: str):
+        """Write file with `content` to `full_path`"""
+        if full_path.exists():
+            with open(full_path, "r") as handle:
+                cached_content = handle.read()
+            if content == cached_content:
+                del cached_content
+                return
+            del cached_content
+        with open(full_path, "w") as handle:
+            handle.write(content)
+
+    optimade_dir = TOP_DIR.joinpath("optimade")
+    docs_dir = TOP_DIR.joinpath("docs/api_reference")
+
+    unwanted_subdirs = ("__pycache__", "data", "grammar", "static")
+
+    pages_template = 'title: "{name}"\n'
+    md_template = "# {name}\n\n::: {py_path}\n"
+    models_template = (
+        md_template + f"{' ' * 4}rendering:\n{' ' * 6}show_if_no_docstring: true\n"
+    )
+
+    if docs_dir.exists() and pre_clean:
+        shutil.rmtree(docs_dir, ignore_errors=True)
+        if docs_dir.exists():
+            raise RuntimeError(f"{docs_dir} should have been removed!")
+    docs_dir.mkdir(exist_ok=True)
+
+    for dirpath, dirnames, filenames in os.walk(optimade_dir):
+        for unwanted_dir in unwanted_subdirs:
+            if unwanted_dir in dirnames:
+                # Avoid walking into or through unwanted directories
+                dirnames.remove(unwanted_dir)
+
+        relpath = Path(dirpath).relative_to(optimade_dir)
+
+        # Create `.pages`
+        if str(relpath) == ".":
+            write_file(
+                full_path=docs_dir.joinpath(".pages"),
+                content=pages_template.format(name="API Reference"),
+            )
+            continue
+
+        docs_sub_dir = docs_dir.joinpath(relpath)
+        docs_sub_dir.mkdir(exist_ok=True)
+        write_file(
+            full_path=docs_sub_dir.joinpath(".pages"),
+            content=pages_template.format(name=str(relpath).split("/")[-1]),
+        )
+
+        # Create markdown files
+        for filename in filenames:
+            if re.match(r".*\.py$", filename) is None or filename == "__init__.py":
+                # Not a Python file: We don't care about it!
+                # Or filename is `__init__.py`: We don't want it!
```

Comment:
Do we not want to capture e.g. module-level docstrings (that would be in the `__init__.py`)?

Reply:
The issue is that it doesn't come off nicely in the API Reference section, since it will "break" the system a little. One will have to have a `__init__` entry in the ToC to work this in. I have tried here to avoid that, but it may be that it's still better than nothing?

Reply:
Yeah I get you, not sure the best way of fixing it really with the current tree structure we have in the docs. Unless it could be renamed to "contents" and include the submodule docstrings... probably more trouble than its worth.

Reply:
Yeah, but not a bad idea. I'll give it a go.
And then make sure to also name the ToC header something non-API-like, e.g., "Contents" as you suggest (the capitalization being important here).

Reply:
So mostly this presents the issue that `mkdocstrings` cannot handle the `__all__` "imports". I.e., it errors when trying to build the docs due to a "NameError":
```sh
    __all__ = exceptions.__all__ + references.__all__ + structures.__all__  # noqa: F405
NameError: name 'exceptions' is not defined
```

Reply:
This is also why I gave up and moved the resource-specific adapters I had written in the `__init__.py` files to new `adapters.py` files, so they could be included.

Reply:
But I could instead try to create a custom markdown file that already exists at the top of each module that can parse in the doc string from the `__init__.py` file? Unfortunately, it won't get anything else from there. Like, e.g., the `validator()` function in `optimade.validator.__init__`

Reply:
So if you want to include this, I suggest instead to move it to a new file, e.g., `cli.py` and keep `__init__.py` only to create an `__all__` if needed.